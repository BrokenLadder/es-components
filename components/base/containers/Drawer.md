Use ``Drawer`` and ``DrawerPanel`` components to wrap content in collapsable panels.

### Basic Example:
```
<Drawer>
  <DrawerPanel header="Collapsable Item 1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 2">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```

### Drawer Options

Option            | Type                   | Description
----------------- | ---------------------- | ---------------------------------------------------------
isAccordion       | boolean                | Only allows one DrawerPanel to be open at a time
defaultActiveKeys | string or string array | Specify which panels are opened by default
closedIconName    | string                 | Override the default plus icon with another OE icon name
openedIconName    | string                 | Override the default minus icon with another OE icon name
className         | string or object       | Add additional CSS classes to the root drawer element

### DrawerPanel Options

Option    | Type             | Description
--------- | ---------------- | ---------------------------------------------------------------
header    | string           | Header text displayed next to the open/close icon
key       | string           | Manually set the key value for the panel (otherwise uses index)
className | string or object | Add additional CSS classes to the drawer item element
noPadding | boolean          | Removes the default padding from the panel body


### Advanced Example:
```
<Drawer isAccordion defaultActiveKeys="1" closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
  <DrawerPanel header="Collapsable Item 1" key="1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 2" key="2" className="alt-color">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 3" key="3" noPadding>
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```
