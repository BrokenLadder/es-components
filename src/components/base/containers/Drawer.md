Use ``Drawer`` and ``Drawer.Panel`` components to wrap content in collapsable panels.

### Basic Example:
```
<Drawer>
  <Drawer.Panel title="Panel 1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 2">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```

### Advanced Example:

Use the ``noPadding`` property to remove default padding within a panel (useful for lists and tables).
You can customize the key property of each Panel; if not specified the default is a zero-based array.
Both the Drawer and Drawer.Panel components will accept additional classNames.

```
<Drawer isAccordion defaultActiveKeys="1" closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
  <Drawer.Panel title="Panel 1" key="1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 2" key="2" noPadding>
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 3" key="3" className="alt-color">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```
