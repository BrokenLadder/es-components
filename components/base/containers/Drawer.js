import React, { PropTypes, Children, Component } from 'react';
import classNames from 'classnames';
import './drawer.less';

function toArray(activeKey) {
  let currentActiveKey = activeKey;
  if (!Array.isArray(currentActiveKey)) {
    currentActiveKey = currentActiveKey ? [currentActiveKey] : [];
  }
  return currentActiveKey;
}

class Drawer extends Component {
  constructor(props) {
    super(props);

    const currentActiveKey = this.props.defaultActiveKeys;
    this.state = {
      activeKey: toArray(currentActiveKey)
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('activeKey' in nextProps) {
      this.setState({
        activeKey: toArray(nextProps.activeKey)
      });
    }
  }

  onClickItem(key) {
    return () => {
      let activeKey = this.state.activeKey;

      if (this.props.isAccordion) {
        activeKey = activeKey[0] === key ? [] : [key];
      } else {
        activeKey = [...activeKey];
        const index = activeKey.indexOf(key);
        const isActive = index > -1;
        if (isActive) {
          activeKey.splice(index, 1);
        } else {
          activeKey.push(key);
        }
      }

      this.setState({ activeKey });
    };
  }

  getItems() {
    const activeKey = this.state.activeKey;
    const { isAccordion } = this.props;
    const newChildren = [];

    Children.forEach(this.props.children, (child, index) => {
      if (!child) return;
      // If there is no key provided, use the panel order as default key
      const key = child.key || String(index);
      const header = child.props.header;

      let isActive = false;
      if (isAccordion) {
        isActive = activeKey[0] === key;
      } else {
        isActive = activeKey.indexOf(key) > -1;
      }

      const props = {
        key,
        header,
        isActive,
        children: child.props.children,
        onItemClick: this.onClickItem(key).bind(this),
        closedIconName: this.props.closedIconName,
        openedIconName: this.props.openedIconName
      };

      newChildren.push(React.cloneElement(child, props));
    });

    return newChildren;
  }

  render() {
    const { className } = this.props;
    const classes = classNames('drawer', className);

    return (
      <div className={classes}>
        {this.getItems()}
      </div>
    );
  }
}

Drawer.propTypes = {
  children: PropTypes.any,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]),
  closedIconName: PropTypes.string,
  defaultActiveKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  isAccordion: PropTypes.bool,
  openedIconName: PropTypes.string
};

Drawer.defaultProps = {
  closedIconName: 'plus',
  isAccordion: false,
  openedIconName: 'minus'
};

export default Drawer;
