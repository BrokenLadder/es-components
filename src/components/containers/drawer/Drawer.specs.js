/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Drawer from './Drawer';

jest.mock('../../util/generateAlphaName', () => () => 'abcdef');

describe('drawer component', () => {
  describe('drawer', () => {
    let instanceToRender;

    beforeEach(() => {
      instanceToRender = (
        <Drawer className="important">
          <Drawer.Panel
            title="collapse 1"
            key="1"
            className="first"
            titleAside="side text"
          >
            first
          </Drawer.Panel>
          <Drawer.Panel title="collapse 2" key="2" className="second" noPadding>
            second
          </Drawer.Panel>
          <Drawer.Panel title="collapse 3" key="3" className="third">
            third
          </Drawer.Panel>
        </Drawer>
      );
    });

    it('renders as expected', () => {
      const tree = renderer.create(instanceToRender).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('click should toggle panel state', () => {
      const drawerInstance = mount(instanceToRender);
      const firstPanel = drawerInstance.find('.first');
      const panelButton = firstPanel.find('button');

      expect(panelButton.prop('aria-expanded')).toBe(false);

      panelButton.simulate('click');
      expect(panelButton.prop('aria-expanded')).toBe(true);

      panelButton.simulate('click');
      expect(panelButton.prop('aria-expanded')).toBe(false);
    });

    it('should allow more than one drawer to be opened at a time', () => {
      const drawerInstance = mount(instanceToRender);
      const firstPanel = drawerInstance.find('.first');
      const firstButton = firstPanel.find('button');
      const thirdPanel = drawerInstance.find('.third');
      const thirdButton = thirdPanel.find('button');

      expect(firstButton.prop('aria-expanded')).toBe(false);
      expect(thirdButton.prop('aria-expanded')).toBe(false);

      firstButton.simulate('click');
      expect(firstButton.prop('aria-expanded')).toBe(true);
      expect(thirdButton.prop('aria-expanded')).toBe(false);

      thirdButton.simulate('click');
      expect(firstButton.prop('aria-expanded')).toBe(true);
      expect(thirdButton.prop('aria-expanded')).toBe(true);
    });

    it('should set multiple default panels open at start', () => {
      const drawerInstance = mount(
        <Drawer defaultActiveKeys={['2', '3']}>
          <Drawer.Panel title="collapse 1" key="1">first</Drawer.Panel>
          <Drawer.Panel title="collapse 2" key="2">second</Drawer.Panel>
          <Drawer.Panel title="collapse 3" key="3">third</Drawer.Panel>
        </Drawer>
      );

      const panels = drawerInstance.find('[aria-expanded]');
      expect(panels.at(0).prop('aria-expanded')).toBe(false);
      expect(panels.at(1).prop('aria-expanded')).toBe(true);
      expect(panels.at(2).prop('aria-expanded')).toBe(true);
    });
  });

  describe('accordion', () => {
    let instanceToRender;

    beforeEach(() => {
      instanceToRender = (
        <Drawer isAccordion defaultActiveKeys="1">
          <Drawer.Panel title="collapse 1" key="1" className="first">
            first
          </Drawer.Panel>
          <Drawer.Panel title="collapse 2" key="2" className="second">
            second
          </Drawer.Panel>
          <Drawer.Panel title="collapse 3" key="3" className="third">
            third
          </Drawer.Panel>
        </Drawer>
      );
    });

    it('renders as expected', () => {
      const tree = renderer.create(instanceToRender).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('default active key should set panel open', () => {
      const accordionInstance = mount(instanceToRender);
      const panels = accordionInstance.find('[aria-expanded]');

      expect(panels.at(0).prop('aria-expanded')).toBe(true);
      expect(panels.at(1).prop('aria-expanded')).toBe(false);
      expect(panels.at(2).prop('aria-expanded')).toBe(false);
    });

    it('should only allow one drawer to be opened at a time', () => {
      const accordionInstance = mount(instanceToRender);
      const firstPanel = accordionInstance.find('.first');
      const firstButton = firstPanel.find('button');
      const thirdPanel = accordionInstance.find('.third');
      const thirdButton = thirdPanel.find('button');

      expect(firstButton.prop('aria-expanded')).toBe(true);
      expect(thirdButton.prop('aria-expanded')).toBe(false);

      thirdButton.simulate('click');
      expect(firstButton.prop('aria-expanded')).toBe(false);
      expect(thirdButton.prop('aria-expanded')).toBe(true);
    });
  });
});
