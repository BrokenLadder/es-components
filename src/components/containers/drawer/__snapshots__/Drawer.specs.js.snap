exports[`drawer component accordion renders as expected 1`] = `
<div
  className="sc-EHOje ktKLsF">
  <div
    className="first sc-bwzfXH dYpfLQ">
    <div
      aria-controls="abcdef"
      aria-expanded={true}
      className="sc-htpNat kITHuH"
      onClick={[Function]}
      role="tab">
      <span>
        <i
          aria-hidden={true}
          className="sc-bxivhb beQLQJ sc-bdVaJa dRgmsV"
          content="63d"
          fontFamily="oecom-icons"
          fontSize="inherit" />
        collapse 1
      </span>
    </div>
    <div
      className="sc-ifAKCX eMUdZt"
      id="abcdef"
      style={
        Object {
          "display": null,
          "height": "auto",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      }>
      <div
        style={
          Object {
            "overflow": "hidden",
          }
        }>
        first
      </div>
    </div>
  </div>
  <div
    className="second sc-bwzfXH dYpfLQ">
    <div
      aria-controls="abcdef"
      aria-expanded={false}
      className="sc-htpNat kITHuH"
      onClick={[Function]}
      role="tab">
      <span>
        <i
          aria-hidden={true}
          className="sc-bxivhb beQLQJ sc-bdVaJa iOqSxE"
          content="63f"
          fontFamily="oecom-icons"
          fontSize="inherit" />
        collapse 2
      </span>
    </div>
    <div
      className="sc-ifAKCX eMUdZt"
      id="abcdef"
      style={
        Object {
          "display": "none",
          "height": "0",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
  <div
    className="third sc-bwzfXH dYpfLQ">
    <div
      aria-controls="abcdef"
      aria-expanded={false}
      className="sc-htpNat kITHuH"
      onClick={[Function]}
      role="tab">
      <span>
        <i
          aria-hidden={true}
          className="sc-bxivhb beQLQJ sc-bdVaJa iOqSxE"
          content="63f"
          fontFamily="oecom-icons"
          fontSize="inherit" />
        collapse 3
      </span>
    </div>
    <div
      className="sc-ifAKCX eMUdZt"
      id="abcdef"
      style={
        Object {
          "display": "none",
          "height": "0",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
</div>
`;

exports[`drawer component drawer renders as expected 1`] = `
<div
  className="important sc-EHOje ktKLsF">
  <div
    className="first sc-bwzfXH dYpfLQ">
    <div
      aria-controls="abcdef"
      aria-expanded={false}
      className="sc-htpNat kITHuH"
      onClick={[Function]}
      role="tab">
      <span>
        <i
          aria-hidden={true}
          className="sc-bxivhb beQLQJ sc-bdVaJa iOqSxE"
          content="63f"
          fontFamily="oecom-icons"
          fontSize="inherit" />
        collapse 1
      </span>
      <span>
        side text
      </span>
    </div>
    <div
      className="sc-ifAKCX eMUdZt"
      id="abcdef"
      style={
        Object {
          "display": "none",
          "height": "0",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
  <div
    className="second sc-bwzfXH dYpfLQ">
    <div
      aria-controls="abcdef"
      aria-expanded={false}
      className="sc-htpNat kITHuH"
      onClick={[Function]}
      role="tab">
      <span>
        <i
          aria-hidden={true}
          className="sc-bxivhb beQLQJ sc-bdVaJa iOqSxE"
          content="63f"
          fontFamily="oecom-icons"
          fontSize="inherit" />
        collapse 2
      </span>
    </div>
    <div
      className="sc-ifAKCX eMUdZt"
      id="abcdef"
      style={
        Object {
          "display": "none",
          "height": "0",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
  <div
    className="third sc-bwzfXH dYpfLQ">
    <div
      aria-controls="abcdef"
      aria-expanded={false}
      className="sc-htpNat kITHuH"
      onClick={[Function]}
      role="tab">
      <span>
        <i
          aria-hidden={true}
          className="sc-bxivhb beQLQJ sc-bdVaJa iOqSxE"
          content="63f"
          fontFamily="oecom-icons"
          fontSize="inherit" />
        collapse 3
      </span>
    </div>
    <div
      className="sc-ifAKCX eMUdZt"
      id="abcdef"
      style={
        Object {
          "display": "none",
          "height": "0",
          "overflow": "hidden",
          "transition": "height .35s ease",
        }
      } />
  </div>
</div>
`;
